File: LeetCode/Java/src/solutions/p2492_MinimumScorePahtBetweenTwoCities/P2492.java
package solutions.p2492_MinimumScorePahtBetweenTwoCities;

import java.util.LinkedList;
import java.util.Queue;

public class P2492 {
    public int minScore(int n, int[][] roads) {
        /*
         * Solution v1: Brute Force
         *
         * Memory Limit Exceeded
         */
        int ans = Integer.MAX_VALUE;

        int[][] path = new int[n + 1][n + 1];
        for (int[] road : roads) {
            path[road[0]][road[1]] = road[2];
            path[road[1]][road[0]] = road[2];
        }
        System.out.println(path);
        boolean[] visited = new boolean[n + 1];
        System.out.println(visited);
        Queue<Integer> queue = new LinkedList<>();
        queue.add(1);

        while (!queue.isEmpty()) {
            int s = queue.poll();
            for (int i = 0; i < n; i++) {
                if (path[s][i] > 0) {
                    if (!visited[i]) {
                        queue.add(i);
                        visited[i] = true;
                    }
                }

                ans = Math.min(ans, path[s][i]);
            }
        }

        return ans;
    }
}

File: LeetCode/Python/solutions/p2492.py
import collections
from typing import List


class Solution:
    def minScore(self, n: int, roads: List[List[int]]) -> int:
        #
        # Solution v1: Brute Force
        # iterate through roads, just ignore the param n
        #
        # Runtime: 1642 ms @ (beats) 92.28%
        # Memory Usage: 76.7 MB @ (beats) 31.88%
        #
        ans = float("inf")

        path = collections.defaultdict(dict)
        for s, e, d in roads:
            path[s][e] = d
            path[e][s] = d

        visited = set()
        q = collections.deque()
        q.append(1)

        while q:
            s = q.popleft()
            for e in path[s]:
                if e not in visited:
                    q.append(e)
                    visited.add(e)
                ans = min(ans, path[s][e])

        return ans


if __name__ == "__main__":
    solution = Solution()

    n = 4
    roads = [[1, 2, 9], [2, 3, 6], [2, 4, 5], [1, 4, 7]]
    print(f"return: {solution.minScore(n, roads)}")  # 5

    n = 4
    roads = [[1, 2, 2], [1, 3, 4], [3, 4, 7]]
    print(f"return: {solution.minScore(n, roads)}")  # 2
